scripts/assemble_changelog.py

```diff
@@ -0,0 +1,457 @@
+        in STANDARD_SECTIONS in assemble_changelog.py or already present
+        in ChangeLog.md. Section titles must match byte-for-byte except that
+        leading or trailing whitespace is ignored.
+        """
+        filename = input_stream.name
+        current_section = None
+        for line_number, line in enumerate(input_stream, 1):
+            if not line.strip():
+                continue
+            level, content = self.title_level(line)
+            if level == 3:
+                current_section = content
+                if current_section not in self.section_content:
+                    raise InputFormatError(filename, line_number,
+                                           'Section {} is not recognized',
+                                           str(current_section)[1:])
+            elif level == 0:
+                if current_section is None:
+                    raise InputFormatError(filename, line_number,
+                                           'Missing section title at the beginning of the file')
+                self.section_content[current_section].append(line)
+            else:
+                raise InputFormatError(filename, line_number,
+                                       'Only level 3 headers (###) are permitted')
+
+    def write(self, filename):
+        """Write the changelog to the specified file.
+        """
+        with open(filename, 'wb') as out:
+            for line in self.header:
+                out.write(line)
+            for section, lines in self.section_content.items():
+                if not lines:
+                    continue
+                out.write(b'### ' + section + b'\n\n')
+                for line in lines:
+                    out.write(line)
+                out.write(b'\n')
+            for line in self.trailer:
+                out.write(line)
+
+
+@functools.total_ordering
+class FileMergeTimestamp:
+    """A timestamp indicating when a file was merged.
+
+    If file1 was merged before file2, then
+    FileMergeTimestamp(file1) <= FileMergeTimestamp(file2).
+    """
+
+    # Categories of files. A lower number is considered older.
+    MERGED = 0
+    COMMITTED = 1
+    LOCAL = 2
+
+    @staticmethod
+    def creation_hash(filename):
+        """Return the git commit id at which the given file was created.
+
+        Return None if the file was never checked into git.
+        """
+        hashes = subprocess.check_output(['git', 'log', '--format=%H', '--', filename])
+        if not hashes:
+            # The file was never checked in.
+            return None
+        hashes = hashes.rstrip(b'\n')
+        last_hash = hashes[hashes.rfind(b'\n')+1:]
+        return last_hash
+
+    @staticmethod
+    def list_merges(some_hash, target, *options):
+        """List merge commits from some_hash to target.
+
+        Pass options to git to select which commits are included.
+        """
+        text = subprocess.check_output(['git', 'rev-list',
```

Comment:
So far this program is only meant to be run manually by Mbed TLS developers. In the foreseeable future, it may be run in a controlled CI environment. So I don't think we need nicer error handling. (In particular, I deliberately didn't cater for running this without being in a git working directory. I should document this though.) If this function catches `CalledProcessError`, what would it do? What details would be useful that are not in the error trace?

Reply:
I think my thoughts were more along the lines of using POpen to capture both stdout / stderr, then if the command fails, dumping stderr and quitting. I agree catching `CalledProcessError` on its own is pretty useless as it only gives the returned error code. It just seems to me that there are a number of reasons why this could fail (as with any usage of subprocess), and it could take time to debug this without some output as to what went wrong, when the script literally just quits after catching CalledProcessError (and the trace back will only show the error code)

Reply:
And we're already getting stderr on stderr.

Reply:
Oh, apologies. I misread the documentation and presumed this function would capture stderr as well as stdout, hence thinking the only thing you would get would be an error number.